### ISSUE TITLE
triggerCharacters cause completions to trigger in unexpected places

### ISSUE THREAD

**Issue by rchl** [sublimelsp/LSP#2090]
**Describe the bug**

Many servers (pyright, typescript-language-server, ...) specify `.` as a trigger character. This is most often unnecessary (as ST already triggers completions on `.`) and it also causes annoyance within strings (for example) when typing `.` as nobody expects completions there.

We have ways of disabling triggerCharacters per server and we even have ways to specify custom scopes for triggers but this issue is more about figuring out some holistic way of handling that issue.

My unverified, of-the-top-the-head thinking is that we could automatically apply ST's scopes from `auto_complete_selector` to every character that the server specifies.

Though there are cases like `typescript-language-server` where it might be useful to trigger completions within JSDoc comments and one is importing types.

**To Reproduce**
Install LSP-pyright, open python file and type `.` within a string.

**Expected behavior**

No completions triggered

**Comment by rchl** [comment]
Hmm, I think `auto_complete_selector` is already respected for those trigger characters but `string.quoted.double` is not excluded by default so it doesn't help.

Default value of `auto_complete_selector` is `meta.tag, source - comment - string.quoted.double.block - string.quoted.single.block - string.unquoted.heredoc`.

**Comment by rchl** [comment]
I don't think there actually exists a generic solution for this, based on what I've realized in the second comment.

Whether it's OK to generally ignore all trigger characters would have to be verified. For example take typescript-language-server, figure out the purpose of each trigger character and check results before and after disabling that trigger.

**Comment by jwortmann** [comment]
It could possibly be a server-specific configuration option whether to ignore trigger characters or not, but they probably shouldn't be ignored by default.

For example the TexLab language server registers `'triggerCharacters': ['\\', '{', '}', '@', '/', ' ']` (e.g. in LaTeX commands always begin with a backslash), and if they were ignored, ST wouldn't show any automatic completions in LaTeX files at all (because the default "auto_complete_selector" doesn't include the `text.tex.latex` base scope).

But maybe the best solution would be to allow users to completely override the "triggerCharacters" (instead of just on/off toggle). I've just checked with LSP-TexLab and I personally find it really annoying that the server also includes the space `' '` in the trigger characters, i.e. the autocompletion popup is basically open the entire time when you type something, even after words in a normal sentence.

---

Btw, I just discovered that LSP's "inhibit_word_completions" setting, which is enabled by default, doesn't have any effect (at least with LSP-TexLab). I still get all the word completions from the buffer when typing a sentence, even though the server returns no items...
```
:: <<< LSP-TexLab 9: {'items': [], 'isIncomplete': False}
```

**Comment by rchl** [comment]
> But maybe the best solution would be to allow users to completely override the "triggerCharacters" (instead of just on/off toggle).

We have scope-based override already which is generally superior to just plain characters.

https://github.com/sublimelsp/LSP-json/blob/a65342f9908b0a759c148a574815d20811201b06/LSP-json.sublime-settings#L15-L21

**Comment by rchl** [comment]
> Btw, I just discovered that LSP's "inhibit_word_completions" setting, which is enabled by default, doesn't have any effect

I think it doesn't have effect in exactly this case - when server returns no completions.

**Comment by rwols** [comment]
They're also disabled here https://github.com/sublimelsp/LSP-lua/blob/df49aee213504a32722ec7f523abb74846c031bb/LSP-lua.sublime-settings#L23-L30

**Comment by jwortmann** [comment]
> Would it work to disable all trigger characters and move entirely to `auto_complete_selector` like LSP-json does? I guess it depends on scopes?

That would work under the assumption that all of the trigger characters have a unique scope, and this scope is not also used for other characters. But for this example from TexLab, it's not possible to match all the trigger characters, for example in
```latex
\include{path/to/file}
```
the `/` don't get any special scope, so completions wouldn't trigger after the slash.

And in general, for this approach I need to manually find all the scope names used by each of the trigger characters, which could be tricky because the scope for a particular character can in general depend on where/how it is used. Doing that may be justified for the purpose of using it as predefined default value in the LSP-\*.sublime-settings file, but this setting is quite subjective, e.g. other users might like the autocompletion after space in LaTeX files. So if users would want to adjust the trigger characters to their personal preference, to do it this way is quite awkward.

> 
> > And this setting also doesn't seem to work reliably. For example it's already configured to exclude comments
> 
> Maybe because original server triggerCharacters are not disabled?

Well, if they were disabled, they wouldn't work anymore... so they are not disabled on purpose I guess. If I understand it correctly, the "auto_complete_selector" should work just as an additional restriction to the trigger characters. And it does seem to work for the `\`  and the other characters within comments, just not for the `{`. So perhaps there is something else going on, or it is indeed a ST bug.

---

And in regard to the opening comment:

> My unverified, of-the-top-the-head thinking is that we could automatically apply ST's scopes from `auto_complete_selector` to every character that the server specifies.

Exactly this seems to be what already happens:
https://github.com/sublimelsp/LSP/blob/e1be509a5fe6c20a7607a79b9aebcb83f8f5541d/plugin/session_view.py#L182-L186

**Comment by jwortmann** [comment]
> > Btw, I just discovered that LSP's "inhibit_word_completions" setting, which is enabled by default, doesn't have any effect
> 
> I think it doesn't have effect in exactly this case - when server returns no completions.

Hm, not sure. I actually have a counter example:

By default, ST often triggers the autocompletions after a space. For example if you start in safe mode and write
```python
import functools
import
```
in Python then it will open the completion popup when you type a space after the second `import` and suggest the `functools` word completion from the buffer.

Personally I find that very annoying, so since the release of ST4 I use the following plugin in my user folder to prevent completions after space, and it works well:
```python
import sublime
import sublime_plugin

class CompletionListener(sublime_plugin.ViewEventListener):

    def on_query_completions(self, prefix, locations):
        # suppress auto complete when typing whitespace
        if not prefix:
            return [], sublime.INHIBIT_WORD_COMPLETIONS | sublime.DYNAMIC_COMPLETIONS
```

So here I don't return any completions and still it prevents ST from showing word completions.

But maybe the bug that LSP's "inhibit_word_completions" doesn't reliably work is caused by using a sublime.CompletionList as the return value? I vaguely remember a request in the ST issue tracker to allow to specify the completion flags early and before the CompletionList gets resolved.

**Comment by rchl** [comment]
> But this variant also works fine and prevents the word completions. So it is not caused just by using the CompletionList, and the bug must either be somewhere in LSP, or there is something else which is affecting this.

I've tested on a simple text document like:

```
\documentclass[12pt]{article}
\usepackage{lingmacros}
\usepackage{tree-dvips}
\begin{document}

\section*{Notes for My Paper}

This is a 
```

with no LSP and your simple plugin also doesn't prevent word completions from showing up. Just note that when testing without LSP you need to explicitly request completions with ctrl+space as they won't be triggered on just pressing space due to no extra trigger character being specified.

**Comment by jwortmann** [comment]
Yeah, it seems that ST behaves different when completions are manually triggered, or due to the "auto_complete_triggers" setting. Apparently the flags from `on_query_completions` are ignored in that case. I guess this is another reason in favor of allowing users to override the "triggerCharacters" from the server to tweak them individually. Because currently LSP's "inhibit_word_completions" setting has no effect when the completion popup is triggered from one of the server's "triggerCharacters", and there seems to be no way to fix that.

---

Just out of interest I've also tried what happens when autocomplete is manually triggered with <kbd>Ctrl</kbd> + <kbd>Space</kbd>, and it still seems to significantly improve the completions from ST (tested in safe mode and with all default settings):

This is what I get without the plugin:

![default](https://user-images.githubusercontent.com/6579999/197362421-2dbba9ce-3d6a-4512-b8e9-924315b5484c.gif)

(where does even this junk come from... those single letters are not in the buffer for potentially being suggested as word completions)

And this is with my plugin (tested with the 2nd variant):

![plugin](https://user-images.githubusercontent.com/6579999/197362439-8eee9d7b-7afd-4f4b-a492-d52b85461835.gif)
